qiskit/transpiler/passes/layout/sabre_layout.py

```diff
@@ -303,7 +303,18 @@ def run(self, dag):
         # `ApplyLayout` transpiler pass (which usually does this step), because we're about to apply
         # the layout and routing together as part of resolving the Sabre result.
         physical_qubits = QuantumRegister(self.coupling_map.size(), "q")
-        mapped_dag = DAGCircuit()
+        num_nodes = 2 * (len(physical_qubits) + dag.num_clbits() + dag.num_vars)
+        num_edges = 0
+        for component in components:
+            component_dag = component.dag
+            size = len(component_dag._multi_graph) - 2 * len(component_dag._wires)
+            num_nodes += size
+            num_edges += component_dag._multi_graph.num_edges()
+
+        mapped_dag = DAGCircuit(
+            node_count_hint=num_nodes,
+            edge_count_hint=num_edges,
+        )
```

Comment:
I'm suspicious that the Python-space time to calculate this stuff is quite possibly similar to the Rust-space time to just grow the vectors.

Reply:
Yeah, I was worried about this too. In practice I think the list is only ever a single list entry here though. The code handles disjoint connectivity, but iirc it was unsound to do the routing split up so we bail return an unrouted circuit before this gets called if the number of components is >1.

Reply:
I'm fine to just leave it anyway - it's probably just going to be a wash at worst, rather than hurting.